File: app/src/main/java/com/bitflaker/lucidsourcekit/notification/NotificationOrderManager.java
package com.bitflaker.lucidsourcekit.notification;

import com.bitflaker.lucidsourcekit.database.notifications.entities.NotificationCategory;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

public class NotificationOrderManager {
    private final static long MIN_NOTIFICATION_FUTURE_TIME = 60 * 1000;
    private final List<NotificationScheduleData> scheduleData;
    private final List<Long> markers;
    private boolean hasNotifications;

    public NotificationOrderManager() {
        scheduleData = new ArrayList<>();
        markers = new ArrayList<>();
        hasNotifications = false;
    }

    public static NotificationOrderManager load(List<NotificationCategory> notificationCategories) {
        NotificationOrderManager notOrdMan = new NotificationOrderManager();
        for (NotificationCategory cat : notificationCategories) {
            if(cat.isEnabled() && cat.getDailyNotificationCount() > 0) {
                notOrdMan.scheduleNotificationTimeSpan(cat.getTimeFrom(), cat.getTimeTo(), cat.getDailyNotificationCount(), cat.getId());
            }
        }
        return notOrdMan;
    }

    public void scheduleNotificationTimeSpan(long timeFrom, long timeTo, int count, String id) {
        if(count > 0) {
            hasNotifications = true;
        }
        insertMarkerIfNonExistent(timeFrom);
        insertMarkerIfNonExistent(timeTo);
        long duration = timeTo - timeFrom;
        long delay = duration / count;
        int itemCounter = 0;
        do {
            long time = timeFrom + (delay * itemCounter) + (delay / 2);
            int idx = scheduleData.size();
            for (int i = 0; i < scheduleData.size(); i++) {
                if(scheduleData.get(i).getTimestamp() > time){
                    idx = i;
                    break;
                }
            }
            scheduleData.add(idx, new NotificationScheduleData(time, id));
            itemCounter++;
        }
        while(itemCounter < count);
    }

    public List<NotificationScheduleData> getScheduledData() {
        List<NotificationScheduleData> finalItems = new ArrayList<>();
        for (int markerIndex = 0; markerIndex < markers.size() - 1; markerIndex++) {
            List<NotificationScheduleData> markerItems = getDataInMarker(markerIndex);
            long markerStart = markers.get(markerIndex);
            long markerEnd = markers.get(markerIndex  + 1);
            long markerDuration = markerEnd - markerStart;
            long inMarkerDelay = markerItems.size() == 0 ? 0 : (markerDuration / markerItems.size());

            for (int i = 0; i < markerItems.size(); i++) {
                long time = markerStart + (inMarkerDelay * i) + (inMarkerDelay / 2);
                finalItems.add(new NotificationScheduleData(time, markerItems.get(i).getId()));
            }
        }
        return finalItems;
    }

    private List<NotificationScheduleData> getDataInMarker(int markerIndex) {
        long markerStart = markers.get(markerIndex);
        long markerEnd = markers.get(markerIndex  + 1);
        boolean isInLastMarker = markers.size() == markerIndex + 1;
        List<NotificationScheduleData> items = new ArrayList<>();
        for (int i = 0; i < scheduleData.size(); i++) {
            long currTS = scheduleData.get(i).getTimestamp();
            if(currTS >= markerStart && currTS < markerEnd || isInLastMarker && (currTS >= markerStart && currTS <= markerEnd)){
                items.add(scheduleData.get(i));
            }
        }
        return items;
    }

    private void insertMarkerIfNonExistent(long marker) {
        if(!markers.contains(marker)){
            int idx = markers.size();
            for (int i = 0; i < markers.size(); i++) {
                if(markers.get(i) > marker){
                    idx = i;
                    break;
                }
            }
            markers.add(idx, marker);
        }
    }

    public NotificationScheduleData getNextNotification() {
        List<NotificationScheduleData> data = getScheduledData();
        if(data.size() == 0) {
            return null;
        }

        Calendar cal = Calendar.getInstance();
        long currTime = cal.get(Calendar.HOUR_OF_DAY) * 60 * 60 * 1000;
        currTime += cal.get(Calendar.MINUTE) * 60 * 1000;
        currTime += cal.get(Calendar.SECOND) * 1000;
        currTime += cal.get(Calendar.MILLISECOND);

        for (int i = 0; i < data.size(); i++) {
            if(data.get(i).getTimestamp() > currTime + MIN_NOTIFICATION_FUTURE_TIME){
                return data.get(i);
            }
        }
        NotificationScheduleData nsd = data.get(0);
        nsd.setNextDay(true);
        return nsd;
    }

    public boolean hasNotifications() {
        return hasNotifications;
    }
}

File: app/src/main/java/com/bitflaker/lucidsourcekit/main/notification/NotificationOrderManager.kt
package com.bitflaker.lucidsourcekit.main.notification

import com.bitflaker.lucidsourcekit.database.notifications.entities.NotificationCategory
import com.bitflaker.lucidsourcekit.utils.Tools
import java.util.Calendar

private const val MIN_NOTIFICATION_FUTURE_TIME = 60L * 1000L

class NotificationOrderManager {
    private val scheduleData: MutableList<NotificationScheduleData> = ArrayList()
    private val markers: MutableList<Long> = ArrayList()
    var hasNotifications = false
        private set

    companion object {
        fun load(notificationCategories: List<NotificationCategory>): NotificationOrderManager {
            val manager = NotificationOrderManager()
            for (category in notificationCategories) {
                if (category.isEnabled && category.dailyNotificationCount > 0) {
                    manager.scheduleNotificationTimeSpan(
                        category.timeFrom,
                        category.timeTo,
                        category.dailyNotificationCount,
                        category.id
                    )
                }
            }
            return manager
        }
    }

    fun scheduleNotificationTimeSpan(timeFrom: Long, timeTo: Long, count: Int, id: String?) {
        hasNotifications = count > 0
        insertMarker(timeFrom)
        insertMarker(timeTo)
        val delay = (timeTo - timeFrom) / count
        for (i in 0..count - 1) {
            val time = timeFrom + (delay * i) + (delay / 2)
            var idx = scheduleData.size
            for (i in scheduleData.indices) {
                if (scheduleData[i].timestamp > time) {
                    idx = i
                    break
                }
            }
            scheduleData.add(idx, NotificationScheduleData(time, id))
        }
    }

    fun getScheduledData(): List<NotificationScheduleData> {
        val finalItems = ArrayList<NotificationScheduleData>()
        for (markerIndex in 0..markers.size - 2) {
            val markerItems = getDataInMarker(markerIndex)
            val markerStart = markers[markerIndex]
            val markerEnd = markers[markerIndex + 1]
            val markerDuration = markerEnd - markerStart
            val inMarkerDelay = if (markerItems.isEmpty()) 0 else (markerDuration / markerItems.size)

            for (i in markerItems.indices) {
                val time = markerStart + (inMarkerDelay * i) + (inMarkerDelay / 2)
                finalItems.add(NotificationScheduleData(time, markerItems[i].id))
            }
        }
        return finalItems
    }

    private fun getDataInMarker(markerIndex: Int): MutableList<NotificationScheduleData> {
        val markerStart = markers[markerIndex]
        val markerEnd = markers[markerIndex + 1]
        val isInLastMarker = markers.size == markerIndex + 1
        val items = ArrayList<NotificationScheduleData>()
        for (i in scheduleData.indices) {
            val currentTime = scheduleData[i].timestamp
            if (currentTime >= markerStart && currentTime < markerEnd || isInLastMarker && (currentTime >= markerStart && currentTime <= markerEnd)) {
                items.add(scheduleData[i])
            }
        }
        return items
    }

    private fun insertMarker(marker: Long) {
        if (markers.contains(marker)) {
            return
        }
        var idx = markers.size
        for (i in markers.indices) {
            if (markers[i] > marker) {
                idx = i
                break
            }
        }
        markers.add(idx, marker)
    }

    fun getNextNotification(): NotificationScheduleData? {
        val data = getScheduledData()
        if (data.isEmpty()) {
            return null
        }

        val currentTime = Tools.getTimeOfDayMillis(Calendar.getInstance())
        for (entry in data) {
            if (entry.timestamp > currentTime + MIN_NOTIFICATION_FUTURE_TIME) {
                return entry
            }
        }

        val nsd = data[0]
        nsd.isNextDay = true
        return nsd
    }
}
